### ISSUE TITLE
Long-running outgoing webhook breaks sending of messages

### ISSUE THREAD

**Issue by winterstefan** [RocketChat/Rocket.Chat#10062]
### Description:

**Short version**
When having a **slow** Outgoing WebHook Integration for `sendMessage`, it'll crash with a timeout. That sometimes leads to a few further errors where messages are not sent properly.

**Environment**
I'm using both the REST & Realtime API to create a custom chat integration to an existing platform. For testing purposes, I fully checked this error with the standard Rocket.Chat UI.

**Details**
- I'm using a `sendMessage` WebHook for sending message information to another project
  - In short, it's a special form of offline notification
  - The WebHook checks several things while making calls to `Subscriptions` / `Messages` model.
  - If the message is suitable (the opponent should be notified), a manual `HTTP` call is triggered to propagate that information to a different project
  - Usually, that (local area) request takes just a few ms. But with high load, the script + request may take >3000ms
- I know the current timeout of [3000ms](https://github.com/RocketChat/Rocket.Chat/blob/develop/packages/rocketchat-integrations/server/lib/triggerHandler.js#L306). That's hardcoded.
- As far as I see (from the code), the hook should be very independent and should not affect the sending of other messages.

### Server Setup Information:
I mainly get this error when using Rocket.Chat with a vagrant VM. In some rare cases, the described error affects a medium-sized kubernetes cluster (3 instances), as well.

* Version of Rocket.Chat Server: 0.62.1 (tested with 0.59.3 as well)
* Operating System: Linux (4.4.0-116-generic)
* Deployment Method(snap/docker/tar/etc): manually (local VM)
* Number of Running Instances: 1
* DB Replicaset Oplog: 
* Node Version: v8.9.4
* mongoDB Version: v3.4.13

### Steps to Reproduce:

1. Create Outgoing WebHook for `sendMessage`, that'll take longer than 3 seconds (in my case: HTTP request)
2. Restart Rocket.Chat server (just for having a clean env); Open logs
3. Open Rocket.Chat UI, create new room (1on1 private chat in my example)
4. Write a message for triggering the webhook
5. Look at the log :-)

_(If I deactivate the WebHook, the error isn't there at all)_

### Expected behavior:

- All sent messages are handled correctly and will be displayed to me and my opponent.

### Actual behavior:

- Sent messages aren't handled correctly
- Different sort of further errors:
  - WebSocket frames doesn't show any outgoing requests to the server
  - Message arrive server, but JSON looks defect (see log below)

### Relevant logs:
- Most obvious log: **Script execution timed out** after 3000ms
```
Mar 06 12:44:06 calida RocketChat[11645]: rocketchat_logger rocketchat_logger.js:278 Integrations ➔ Outgoing WebHook.error   Error: Script execution timed out.
Mar 06 12:44:06 calida RocketChat[11645]:       at ContextifyScript.Script.runInContext (vm.js:59:29)
Mar 06 12:44:06 calida RocketChat[11645]:       at ContextifyScript.Script.runInNewContext (vm.js:65:15)
Mar 06 12:44:06 calida RocketChat[11645]:       at Object.runInNewContext (vm.js:135:38)
Mar 06 12:44:06 calida RocketChat[11645]:       at RocketChatIntegrationHandler.executeScript (/var/lib/rocket.chat/bundle/programs/server/packages/rocketchat_integrations.js:731:30)
Mar 06 12:44:06 calida RocketChat[11645]:       at RocketChatIntegrationHandler.executeTriggerUrl (/var/lib/rocket.chat/bundle/programs/server/packages/rocketchat_integrations.js:1115:19)
Mar 06 12:44:06 calida RocketChat[11645]:       at RocketChatIntegrationHandler.executeTrigger (/var/lib/rocket.chat/bundle/programs/server/packages/rocketchat_integrations.js:1025:12)
Mar 06 12:44:06 calida RocketChat[11645]:       at RocketChatIntegrationHandler.executeTriggers (/var/lib/rocket.chat/bundle/programs/server/packages/rocketchat_integrations.js:1018:14)
Mar 06 12:44:06 calida RocketChat[11645]:       at _wrapperFunction (/var/lib/rocket.chat/bundle/programs/server/packages/rocketchat_integrations.js:2565:51)
Mar 06 12:44:06 calida RocketChat[11645]:       at /var/lib/rocket.chat/bundle/programs/server/packages/rocketchat_lib.js:1162:30
Mar 06 12:44:06 calida RocketChat[11645]:       at Array.reduce (<anonymous>:null:null)
```
- Error where the message JSON seems to be **totally** broken.
```
[34mI20180307-10:09:30.536(0) Discarding message with invalid JSON :"stream-notify-room","9rams":["3h5LawnS7Jm2kQxyZ/typing","GuidoVanRossum",true]}{{"msg":"methtd","id":"28","method" 
[34mI20180307-10:09:36.072(0) Discarding message with invalid JSON :"stream-notify-room",33arams":["3h5LawnS7Jm2kQxyZ/typing","GuidoVanRossum",true]}{"msg":"method","id":"253","method": 
[34mI20180307-10:09:29.913(0) Discarding message with invalid JSON {"msg":"method","id":"28sum",true]}{"msg":"method","id":"234","method":"readMessages","params":["3h5LawnS7Jm2kQxyZ"]}{ 
```
- Once I got a duplicate key error. But that _may_ be unrelated / some other error.
```
Mar 08 13:06:44 calida RocketChat[27939]: Exception while invoking method 'sendMessage' WriteError({"code":11000,"index":0,"errmsg":"E11000 duplicate key error collection: rocketchat.rocketchat_message index: _id_ dup key: { : \"oy5HL3ZTcHcLD76gu\" }","op":{"_id":"oy5HL3ZTcHcLD76gu","rid":"Frv7BoZmepnfsJoAy","msg":"zrtz","ts":"2018-03-08T13:06:44.440Z","u":{"_id":"w6fSjr4mLYZ9HoqWP","username":"customer100","name":"customer100"},"unread":true,"mentions":[],"channels":[],"_updatedAt":"2018-03-08T13:06:44.449Z"}})
```

### Questions:
- Can you tell me why one failing WebHook execution affects other executions of that WebHook? AFAIK the hook will be compiled just once, but runs in a separate `vm` context on every call.
- Is that possibly somehow related to #9926?
- Do you have any tips on how to write a WebHook? Any pitfalls to avoid, is it very ugly to make an external API call (well, most integrations are based on that^^)?

**Comment by theorenck** [comment]
Thanks for report it to us. @sampaiodiego can you help  @winterstefan with his doubts?

**Comment by Hudell** [comment]
I've managed to confirm the issue exists and to consistently trigger the error, but it takes more than a single message to cause it. Sometimes it happens after two or three messages, other times it takes more than 30 slow webhook calls to trigger the issue.

Once it happens, the whole server stops working and you have to kill and restart it to use rocket.chat again.

**Comment by winterstefan** [comment]
Thank you very much for taking a deep look at this!

It was a bit tricky to describe the cause, but your explanation seems very applicable. 

May I ask what kind of error message you got - was it the broken JSON?

**Comment by Hudell** [comment]
Most times there's no error shown, the server just stops working. With the debugger attached I get a SIGSEGV.

I tried to remove the script "cache", to make the app recompile the script with every call. This [change](https://github.com/Hudell/Rocket.Chat/tree/removing-script-cache) seems to have made the problem harder to occur, but it may be just a coincidence.

**Comment by graywolf336** [comment]
Quite honestly, this is not directly correlated to the integrations. This currently happens anytime the server goes slow (too many examples to list while the integrations are the easiest to demonstrate) and the client tries to send a message that was already received and inserted into the database. As a result, this happens because the web client provides it's own id to the message and when it failed to receive an "okay" response from the server it tries to send it again.

**Comment by winterstefan** [comment]
Ah, I understand. tyvm @graywolf336 

> (...) this happens because the web client provides (...)

May I ask if it should only affect the **standard Rocket.Chat web client**? Indeed, I did manage to reproduce that error by using the Rocket.Chat UI. But most of the cases happened by using the self-made integration with REST / Realtime API (e.g. Realtime -> `sendMessage`). Then, the WebSocket commonly showed a lot of new outgoing frames, but none of them have been acknowledged by the server.

> This currently happens anytime the server goes slow

Can you think of any suggestion for improving the mentioned WebHook? As I understand, the majority of existing Rocket.Chat integrations (e.g. JIRA, Jenkins, Giphy) depend on HTTP calls*.
I surely can try to improve the performance of the called endpoint. But as I understand, for several reasons the WebHook temporarily slow in a production environment anyways.

*) only difference: Most Outgoing WebHook integrations I looked at used the standard way (`prepare_outgoing_request`). I used a dedicated call via the `HTTP` helper (e.g. because I need custom request headers / options).

> In some rare cases, the described error affects a medium-sized kubernetes cluster (3 instances), as well.

In these example cases, there wasn't any other special load on the environment. I just tested with one single chat user and wrote about 30 messages. Somehow (e.g. maybe when releasing & rebuilding pods) some WebHook's HTTP requests take uncommonly long and (IMHO) trigger the error.

----
 
**TL;DR;**
- Is there any way of 'tuning' / improving
  - my WebHook?
That hook is called per message, fetches the subscriptions of any other room member, fetches the message object (*optional*) and does a custom HTTP call (in some conditions, not for every message)
  - the API integration? e.g. in addition to the generic spam prevention of Rocket.Chat -> having a custom one where disallow even less messages per interval (that for sure wouldn't be a root cause fix^^)
- Since you said it's a problem that can happen for several causes: Could there be an improvement / a fix for Rocket.Chat itself (I know, tricky question 😇 )?

**Comment by sampaiodiego** [comment]
from what I saw from @Hudell tests, it was like described from @winterstefan , a single local instance with no load and only the slow webhook. 

So I would think that the issue is indeed with our webhook scripts `vm`. not sure how we can deal with it currently though 😞

**Comment by winterstefan** [comment]
@sampaiodiego Full ACK on your point, just one quick remark:

> a single local instance with no load and only the slow webhook.

Yep, that was my primary use case. But I (unfortunately) had that issue (twice, IIRC) on a medium sized 3 node system (with no load either), too. 😇 

_(so not only a problem when running Rocket.Chat on a dev VM locally)_

**Comment by winterstefan** [comment]
> not sure how we can deal with it currently though 😞

I fully understand, that this is a very tricky issue. Nevertheless, do you have any thought of a possible fix, yet / a schedule when the issue **might be** addressed?

_Would any additional information help you?_

For now, my only approach would be to deactivate the message-dependent WebHook.
